irohad/ametsuchi/impl/postgres_burrow_storage.cpp

```diff
@@ -0,0 +1,163 @@
+    std::string_view address) {
+  try {
+    int check = 0;
+    sql_ << "delete from burrow_account_key_value "
+            "where address = lower(:address); "
+            "delete from burrow_account_data "
+            "where address = lower(:address) "
+            "returning 1",
+        soci::use(address, "address"), soci::into(check);
+    if (check == 0) {
+      return makeError("account deletion failed");
+    }
+    return Value<void>{};
+  } catch (std::exception const &e) {
+    return makeError(e.what());
+  }
+}
+
+Result<std::optional<std::string>, std::string>
+PostgresBurrowStorage::getStorage(std::string_view address,
+                                  std::string_view key) {
+  try {
+    std::optional<std::string> value;
+    sql_ << "select value from burrow_account_key_value "
+            "where address = lower(:address) and key = lower(:key) ",
+        soci::use(address, "address"), soci::use(key, "key"), soci::into(value);
+    return value;
+  } catch (std::exception const &e) {
+    return makeError(e.what());
+  }
+}
+
+Result<void, std::string> PostgresBurrowStorage::setStorage(
+    std::string_view address, std::string_view key, std::string_view value) {
+  try {
+    int check = 0;
+    sql_ << "insert into burrow_account_key_value (address, key, value) "
+            "values (lower(:address), lower(:key), :value) "
+            "on conflict (address, key) do update set value = excluded.value "
+            "returning 1",
+        soci::use(address, "address"), soci::use(key, "key"),
+        soci::use(value, "value"), soci::into(check);
+    if (check == 0) {
+      return makeError("account key-value storage update failed");
+    }
+    return Value<void>{};
+  } catch (std::exception const &e) {
+    return makeError(e.what());
+  }
+}
+
+Result<void, std::string> PostgresBurrowStorage::storeTxReceipt(
+    std::string_view address,
+    std::string_view data,
+    std::vector<std::string_view> topics) {
+  try {
+    std::optional<size_t> log_idx;
+
+    if (call_id_cache_) {
+      sql_ << "insert into burrow_tx_logs (call_id, address, data) "
+              "values (:call_id, lower(:address), :data) "
+              "returning log_idx",
+          soci::use(call_id_cache_.value(), "call_id"),
+          soci::use(address, "address"), soci::use(data, "data"),
+          soci::into(log_idx);
+    } else {
+      sql_ << "with inserted_call_id as "
+              "("
+              "  insert into engine_calls (tx_hash, cmd_index)"
+              "  values (:tx_hash, :cmd_index)"
+              "  on conflict (tx_hash, cmd_index) do nothing"
+              "  returning call_id"
+              ")"
+              "insert into burrow_tx_logs (call_id, address, data) "
+              "select call_id, :address, :data from "
+              "("
+              "  ("
+              "    select * from inserted_call_id"
+              "  ) union ("
+              "    select call_id from engine_calls"
```

Comment:
Probably this union will not be required, as call id can be inserted once before command execution, and it will be unique for (tx_hash, cmd_idx) pair